src/Symfony/Component/Translation/MessageCatalogue.php

```diff
@@ -232,7 +232,7 @@ public function getResources()
      */
     public function addResource(ResourceInterface $resource)
     {
-        $this->resources[$resource->__toString()] = $resource;
+        $this->resources[(string) $resource] = $resource;
```

Comment:
There is no general approach. I'd say: use the cast when unsure about the type, the call when sure about type+method.